odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1160 @@
+            ``field`` is also this functional's `Functional.range`.
+
+        Returns
+        -------
+        mul : `Functional`
+            Multiplication result.
+
+            If ``other`` is an `Operator`, ``mul`` is a
+            `FunctionalComp`.
+
+            If ``other`` is a scalar, ``mul`` is a
+            `FunctionalRightScalarMult`.
+
+            If ``other`` is a vector, ``mul`` is a
+            `FunctionalRightVectorMult`.
+
+        See Also
+        --------
+        Operator.__mul__ : implementation of __mul__ for Operators.
+        """
+        if isinstance(other, Operator):
+            return FunctionalComp(self, other)
+        elif isinstance(other, Number):
+            # Left multiplication is more efficient, so we can use this in the
+            # case of linear functional.
+            if self.is_linear:
+                return FunctionalLeftScalarMult(self, other)
+            else:
+                return FunctionalRightScalarMult(self, other)
+        elif isinstance(other, LinearSpaceElement):
+            return FunctionalRightVectorMult(self, other)
+        else:
+            return super().__mul__(other)
+
+    def __rmul__(self, other):
+        """Return ``other * self``.
+
+        If ``other`` is an `Operator`, since a functional is also an operator
+        this corresponds to operator composition:
+
+            ``op * func == (x --> op(func(x))``
+
+        If ``other`` is a scalar, this corresponds to left multiplication of
+        scalars with functionals:
+
+            ``scalar * func == (x --> scalar * func(x))``
+
+        If ``other`` is a vector,  since a functional is also an operator this
+        corresponds to left multiplication of vectors with operators:
+
+            ``vector * func == (x --> vector * func(x))``
+
+        Note that left and right multiplications are generally
+        different.
+
+        Parameters
+        ----------
+        other : `Operator`, `LinearSpaceElement` or scalar
+            `Operator`:
+            The `Operator.domain` of ``other`` must match this functional's
+            `Functional.range`.
+
+            `LinearSpaceElement`:
+            ``other`` must be an element of this functionals's
+            `Functional.range`.
+
+            scalar:
+            The `Operator.domain` of this operator must be a
+            `LinearSpace` and ``other`` must be an
+            element of the ``field`` of this operator's
+            `Operator.domain`.
+
+        Returns
+        -------
+        rmul : `Functional` or `Operator`
+            Multiplication result.
+
+            If ``other`` is an `Operator`, ``rmul`` is an `OperatorComp`.
+
+            If ``other`` is a scalar, ``rmul`` is a
+            `FunctionalLeftScalarMult`.
+
+            If ``other`` is a vector, ``rmul`` is a
+            `OperatorLeftVectorMult`.
+        """
+        if other in self.domain.field:
+            if other == 0:
```

Comment:
I guess you are right, but I would postpone such intelligence for later. Testing this is already quite complicated and with these sneaky special cases happening we're gonna have to be careful.